### ISSUE TITLE
Numeric value lost during division of two units

### ISSUE THREAD

**Issue by Famondir** [maths/moodle-qtype_stack#919]
HI all,
I have found a rather strange behaviour. I have two variables of type unit. When I devide those and compare the result in a response tree I get the following error message:
`Your answer needs to be a number together with units. Your answer only has units.`

But when I print the variable it has a numeric value:
<img width="462" alt="image" src="https://user-images.githubusercontent.com/43599267/214330237-4879669c-cb73-4e4f-b750-553a0ce68574.png">

Right now I have to do the following to come around that issue:
Instead of `Sans5: Sans4/Sans3;` I have to calculate:
```
Sans5: su(stack_units_nums(stack_unit_si_to_si_base(Sans4))/stack_units_nums(stack_unit_si_to_si_base(Sans3)), stack_units_units(stack_unit_si_to_si_base(Sans4))/stack_units_units(stack_unit_si_to_si_base(Sans3)));
```

Any ideas how that can happen and how to solve such issues?

**Comment by sangwinc** [comment]
Yes, this is part of the design actually.  Dimensional numerical quantities are inert functions "stack_units".
https://docs.stack-assessment.org/en/Authoring/Units/#dealing-with-units-in-maxima-functions-eg-question-variables-and-prts

In this case it is quite awkward to do what you might, quite reasonably, expect would be a matter of just  `Sans4/Sans3;`.  As you can see from your proposed solution, there is more to division of dimensional quantities than just division - you have to convert both to base units.

I think we could improve the design here, for example, having utility functions such as something like `stack_units_calc(op, u1, u2)` which performs the operation on dimensional quantities.   E.g. in your case `Sans5: stack_units_calc("/", Sans4, Sans3)`.  Would this help?  I'm very happy to add utility functions where there is a genuine need.

Chris

**Comment by Famondir** [comment]
That would be helpful. Right now I can't decide if I'ld prefer `stack_units_calc(op, u1, u2)` or a collection of functions like `stack_units_add(u1, u2)`, `stack_units_mul(u1, u2)` and so on.

Operator overloading is still not possible in PHP 8, is it? Is there a possibility to do operator overloading while parsing a statement like `stack_units_calc("u1*u2+3*u3")`?

Because of missing possibility to do math with units I use numerics instead and have to keep track of the units seperately.

**Comment by aharjula** [comment]
Well, this gets inconvenient fast... Do note that the existence of the `stackunits` wrapper might cause confusion when used where it is not needed. It is needed for these things:
 1. Passing 0 valued things with units to inputs and tests. And everywhere else as well. 
 2. Ensuring sensible display.
 
That first is important as it can stop the question from becoming non-sensical when some randomisation leads to something becoming 0 and the answer then loses units.

Before the whole wrapper existed it was perfectly possible to do the calculations like this:
```
width: 5*cm;
height: 4*m;
depth: 7*mm;
density: 700000*g/(m^3);
mass: density*width*depth*height;
/* Problem here is that the units are not eliminated so lets convert them to base units so they can eliminate. */
mass: stack_unit_si_to_si_base(mass);
/* Unfortunately that is not a `stackunits` object so lets turn it to such. */
mass_su: stackunits_make(mass);
/* Now the would be in kg as we did the conversion to base, so if we want we can ask for it to be turned to something else will pick the unit that gives a reasonable numerical value from the list given and turns things to `stackunits`. */
mass_sensible: stack_unit_si_present(mass,[mg,cg,ng]);
```
And yes it is still perfectly possible to do this, just the last two commands are things that came in with the `stackunits`.

And no PHP has no role in this. All operations happen in Maxima, so if we do "operator overriding," we do it there, and over there, it is called defining simplification rules if we really want to have such, on the other hand having functions that do this could keep some things cleaner.

However, when acting with student answers, the answer is already a `stackunits` object, so doing maths with it becomes difficult unless you convert it back to raw expressions like those used in the example. I don't think there is a convenience function for that conversion and even if there were losing the units in the case of 0 values might be a real problem.

**Comment by jonaslache** [comment]
I found this issue when I was looking for a solution to a problem I had when I used scientific units in a STACK question. In my case I was wondering why in this example `s1` has the value `stackunits(300,m)` as expected

```
t1: stackunits(30,s);
v1: stackunits(10,m/s);
s1: v1*t1;
```

while in the following example both `a1` and `a2` do not have the value `stackunits(0.5,m/s^2)` as I would expect:

```
m1: stackunits(1000,kg);
F1: stackunits(500,N);
F2: stack_unit_si_to_si_base(F1);
a1: F1/m1;
a2: F2/m1;
```

Instead, `a1` has the value `stackunits(500,N)/stackunits(1000,kg)` (displayed as ${\frac{500\ \mathrm{N}}{1000\ \mathrm{k}\mathrm{g}}}$) and `a2` has the value `stackunits(500,(kg*m)/s^2)/stackunits(1000,kg)` (displayed as ${\frac{500 \frac{\mathrm{k}\mathrm{g}\cdot \mathrm{m}}{\mathrm{s}^2}}{1000\ \mathrm{k}\mathrm{g}}}$). Applying `stack_unit_si_to_si_base` to `a1` or `a2` does not help to get the expected result.

Then I read @aharjula's comment from February 2023. If I understand correctly, I can use the following approach - and it does actually work (`a1` has the value `stackunits(1/2,m/s^2)`):

```
F1: 500*N;
m1: 1000*kg;
F2: stackunits_make(F1);
m2: stackunits_make(m1);
a1: stackunits_make(stack_unit_si_to_si_base(F1)/m1);
```
Is this the currently recommended approach when calculating with expressions with units inside a STACK question? If so, would it be worth to add an example to the docs?

Thanks and best regards<br>Jonas
